libnmstate/metadata.py

```diff
@@ -181,3 +193,73 @@ def _generate_route_metadata(desired_state):
                 iface_state[Interface.IPV6][ROUTES].append(route)
             else:
                 iface_state[Interface.IPV4][ROUTES].append(route)
+
+
+def _generate_dns_metadata(desired_state, current_state):
+    """
+    Save DNS configuration to choose interface as metadata.
+    """
+    # Only support up to 2 name servers.
+    servers = desired_state.config_dns.get(DNS.SERVER, [])[:2]
+    searches = desired_state.config_dns.get(DNS.SEARCH, [])
+    # Some interfaces might old DNS configuration which is not included
+    # in desired_state, we need to add them into desired_state in order to
+    # clear its DNS configurations
+    client = nmclient.client()
+    old_dns_ifaces = nm_dns.get_dns_config_iface_names(
+            nm_ipv4.acs_and_ip_profiles(client),
+            nm_ipv6.acs_and_ip_profiles(client)
+    )
+    _include_name_only_iface_state(
+        desired_state, current_state, old_dns_ifaces)
+    if not servers and not searches:
+        return
+    # The route need to be merged before checking default gateways
+    desired_state.merge_route_config(current_state)
+    ipv4_iface, ipv6_iface = nm_dns.find_interfaces_for_name_servers(
+        desired_state.config_iface_routes)
+
+    _save_dns_metadata(desired_state, current_state, ipv4_iface, ipv6_iface,
+                       servers, searches)
+
+
+def _save_dns_metadata(desired_state, current_state, ipv4_iface, ipv6_iface,
+                       servers, searches):
+    index = 0
+    searches_saved = False
+    for server in servers:
+        iface_name = None
+        if iplib.is_ipv6_address(server):
+            iface_name = ipv6_iface
+            family = Interface.IPV6
+        else:
+            iface_name = ipv4_iface
+            family = Interface.IPV4
+        if not iface_name:
+            raise NmstateValueError(
+                'Failed to find suitable interface for saving DNS '
+                'name servers: %s' % server)
+
+        _include_name_only_iface_state(
```

Comment:
We redo here the same step done in line 213 for existing dns interfaces. So I guess maybe we can remove the one from 213? Not sure.

How about we use here several passes to make it both simpler and readable:
```
ifaces_with_family = _list_dns_ifaces(servers, ipv4_iface, ipv6_iface)
_include_name_only_iface_state(desired_state, current_state, [ipv4_iface, ipv6_iface])
for index, (iface, family) in enumerate(ifaces_with_family):
    ...
```